doc/guides/dts/howtos.rst

```diff
@@ -1,479 +1,546 @@
+The build system prints the output file locations:
 
-Common devicetree definitions for K6X are stored in :zephyr_file:`nxp_k6x.dtsi
-<dts/arm/nxp/nxp_k6x.dtsi>`, which is included by both board :file:`.dts`
-files. :zephyr_file:`nxp_k6x.dtsi<dts/arm/nxp/nxp_k6x.dtsi>` in turn includes
-:zephyr_file:`armv7-m.dtsi<dts/arm/armv7-m.dtsi>`, which has common definitions
-for Arm v7-M cores.
+.. code-block:: none
 
-Since :zephyr_file:`nxp_k6x.dtsi<dts/arm/nxp/nxp_k6x.dtsi>` is meant to be
-generic across K6X-based boards, it leaves many devices disabled by default
-using ``status`` properties.  For example, there is a CAN controller defined as
-follows (with unimportant parts skipped):
+   -- Found BOARD.dts: .../zephyr/boards/arm/qemu_cortex_m3/qemu_cortex_m3.dts
+   -- Generated zephyr.dts: .../zephyr/build/zephyr/zephyr.dts
+   -- Generated devicetree_unfixed.h: .../zephyr/build/zephyr/include/generated/devicetree_unfixed.h
 
-.. code-block:: none
+Change ``qemu_cortex_m3`` to the board you are using, of course.
 
-   can0: can@40024000 {
-   	...
-   	status = "disabled";
-   	...
-   };
+.. _dt-get-device:
 
-It is up to the board :file:`.dts` or application overlay files to enable these
-devices as desired, by setting ``status = "okay"``. The board :file:`.dts`
-files are also responsible for any board-specific configuration of the device,
-such as adding nodes for on-board sensors, LEDs, buttons, etc.
+Get a struct device from a devicetree node
+******************************************
 
-For example, FRDM-K64 (but not Hexiwear K64) :file:`.dts` enables the CAN
-controller and sets the bus speed:
+When writing Zephyr applications, you'll often want to get a driver-level
+:ref:`struct device <device_model_api>` corresponding to a devicetree node.
 
-.. code-block:: none
+For example, with this devicetree fragment, you might want the struct device
+for ``serial@40002000``:
+
+.. code-block:: DTS
 
-   &can0 {
-   	status = "okay";
-   	bus-speed = <125000>;
+   / {
+           soc {
+                   serial0: serial@40002000 {
+                           status = "okay";
+                           current-speed = <115200>;
+                           /* ... */
+                   };
+           };
+
+           aliases {
+                   my-serial = &serial0;
+           };
+
+           chosen {
+                   zephyr,console = &serial0;
+           };
    };
 
-The ``&can0 { ... };`` syntax adds/overrides properties on the node with label
-``can0``, i.e. the ``can@4002400`` node defined in the :file:`.dtsi` file.
+Start by making a :ref:`node identifier <dt-node-identifiers>` for the device
+you are interested in. There are different ways to do this; pick whichever one
+works best for your requirements. Here are some examples:
 
-Other examples of board-specific customization is pointing properties in
-``aliases`` and ``chosen`` to the right nodes (see :ref:`dt-alias-chosen`), and
-making GPIO/pinmux assignments.
+.. code-block::
```

Comment:
This causes a missing argument diagnostic during build, needs:

    .. code-block:: c

Reply:
Interesting; I didn't get one. Will add anyway.